FirebaseMLModelDownloader/Sources/CustomModel.swift

```diff
@@ -0,0 +1,44 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+public enum CustomModelFormat {
+  case Unknown
+  case TFLite
+  case TorchScript
+  case CoreML
+}
+
+public struct CustomModel {
+
+  public let modelName: String
+  public internal(set) var modelSize: Int?
+  public internal(set) var modelPath: String?
+  public internal(set) var modelHash: String?
+  public internal(set) var modelFormat: CustomModelFormat?
```

Comment:
I curious if a model with `modelPath == nil` and `modelSize != nil` is a valid one? To reduce amount of inconsistent states we may want to group properties together into a separate struct e.g.:
```
struct CustomModel {
  struct Metadata {
    var modelSize: Int
    var modelPath: String
    ...
  }

  let name: String
  var metadata: Metadata?
}
```

Reply:
My main point is that if we define the model so that it cannot be in an invalid state then compiler will catch some dummy issues and unhandled cases for us. It's relevant for both the internal implementation and the API clients.

So if the struct may represent either a loaded or not-loaded-yet model, then a more expressive API could be something like

```
struct CustomModel {
  struct Metadata {
    var size: Int
    ...
  }
  enum State {
    case initial
    case downloaded(metadata: Metadata)
    ...
  }

  let name: String
  var state: State
}
```

From the other side, the more Swift specific features we use, the more it will be different from the Objective-C API. For Objective-C we will have to declare `NSObject` subclasses instead of structs and will have to transfer the enum with associated type to Obj-C types.

Sorry if I'm missing any ML SDK specifics, I'm just trying to take the most from Swift power since we have access to it 😄

Reply:
We want CustomModel (and ModelDownloader) to be visible from objc, so is it better to have them be subclasses of NSObject in Swift so that we can use the @objc decorator?

Reply:
Alternatively an additional Objective-C specific model/wrapper may be used, so we can fully laverage Swift advantages and provide a nice Swift API. I would prefer to provide nice Swift API as majority of our customers use Firebase from Swift.

@paulb777 @ryanwilson if you have an opinion on it?

Reply:
I agree with the sentiment to optimize for a nice Swift API while providing a sufficient Objective C API.

Reply:
It looks like the `CustomModel` type isn't used in any of the APIs requesting things (I believe it was used before) so likely the modelName can be readonly as well and all the other properties can be non-optional?

Reply:
I decided to go with having the model object always filled out. So there are no optional properties anymore.